### PR TITLE
REGRESSION (iOS 16): <img> with video source fails to render—affects iCloud live photos (IOKit blocking)

### PR DESCRIPTION
**Pull request by jyavenard**
#### 53f533d03c6eb2a4e40ed920a54266d149b4fa39
<pre>
REGRESSION (iOS 16): &lt;img &gt; with video source fails to render—affects iCloud live photos (IOKit blocking)
<a href="https://bugs.webkit.org/show_bug.cgi?id=241589">https://bugs.webkit.org/show_bug.cgi?id=241589</a>

Reviewed by Simon Fraser.

Stop using IOKit in the content process. Instead we create a SharedMemory backed
ShareableBitmap and send it to the content process. The content process takes
ownership of this ShareableBitmap.

Covered by new media/media-hevc-video-as-img.html test

* LayoutTests/media/content/bipbop-hevc.mp4: Added.
* LayoutTests/media/content/test-video-as-img.png: Added.
* LayoutTests/media/media-hevc-video-as-img-expected.html: Added.
* LayoutTests/media/media-hevc-video-as-img.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* Source/WebKit/GPUProcess/media/RemoteImageDecoderAVFProxy.cpp:
(WebKit::RemoteImageDecoderAVFProxy::createFrameImageAtIndex):
* Source/WebKit/GPUProcess/media/RemoteImageDecoderAVFProxy.h:
* Source/WebKit/GPUProcess/media/RemoteImageDecoderAVFProxy.messages.in:
* Source/WebKit/WebProcess/GPU/media/RemoteImageDecoderAVF.cpp:
(WebKit::RemoteImageDecoderAVF::createFrameImageAtIndex):

Canonical link: <a href="https://commits.webkit.org/251557@main">https://commits.webkit.org/251557@main</a>
git-svn-id: <a href="https://svn.webkit.org/repository/webkit/trunk@295552">https://svn.webkit.org/repository/webkit/trunk@295552</a> 268f45cc-cd09-0410-ab3c-d52691b4dbfc
</pre>

### CODE REVIEW

**Comment by smfr** [comment]
This seems like a reasonable fix, but is there a version where we don't copy the video frame back to the webcontent process at all (since it's only going to get sent back to the GPU process)?

**Comment by jyavenard** [comment]
> This seems like a reasonable fix, but is there a version where we don't copy the video frame back to the webcontent process at all (since it's only going to get sent back to the GPU process)?

I'm happy to take this as a follow-up, this is a blocker. Seeing that this is a non-standard feature, at least we restore this feature.

**Comment by jyavenard** [comment]
IIUI, the drawing seems to be happening in the content process using a GraphicContextCG and as such we have to return a CGImagePtr , to start we would have to enable DisplayListDrawing flag to not take the rendering path.

**Comment by smfr** [comment]
> IIUI, the drawing seems to be happening in the content process using a GraphicContextCG and as such we have to return a CGImagePtr , to start we would have to enable DisplayListDrawing flag to not take the rendering path.

If you were looking at macOS, right. If you look at iOS, where DOM Rendering in GPU Process is enabled, I think it would all happen in the GPU Process.

File: LayoutTests/media/media-hevc-video-as-img.html [hunk 1]
```diff
@@ -0,0 +1,9 @@
+<html>
+<head>
+    <meta name="fuzzy" content="maxDifference=0-84; totalPixels=0-293510" />
```
Comment (smfr) [inline LayoutTests/media/media-hevc-video-as-img.html:3]: That's a lot of fuzziness. Why so many pixels?

**Review by smfr — APPROVED** [review]

File: LayoutTests/media/media-hevc-video-as-img.html (hunk 1 above)
Comment (jyavenard) [reply LayoutTests/media/media-hevc-video-as-img.html:3]: > That's a lot of fuzziness. Why so many pixels?

Each decoder on each platform produce a different output with hevc content. The png was produced using ffmpeg software decoder. 
The aim of the test really is to make sure we have an image out. So fuzziness doesn't really matter too much so long thst the difference is overall small

**Comment by jyavenard** [comment]
> > IIUI, the drawing seems to be happening in the content process using a GraphicContextCG and as such we have to return a CGImagePtr , to start we would have to enable DisplayListDrawing flag to not take the rendering path.
> 
> If you were looking at macOS, right. If you look at iOS, where DOM Rendering in GPU Process is enabled, I think it would all happen in the GPU Process.

I did investigate this first as it was the suggested approach by Youenn, however there's quite a bit to do to allow for proxied CGImagePtr to refer to a other process.
First we will have to use the Recorder graphic context, then figure out how caching will work, along lifetime of the image. 
There's also the issue of ownership, which process gets blamed here?

It's an extensive change I think, and on a code path I'm not familiar with.

**Pull request merged**